### PR TITLE
Hive: Make lock check retries backoff exponentially

### PR DESCRIPTION
**Pull request by raptond**
50 milliseconds (constant) sleep time between "checking lock status" thrashes hive metastore databases when multiple jobs try to commit to the same Iceberg table. This fix allows the frequency of "checking the WAITING lock status" configurable and makes use of Tasks to backoff exponentially.

Every time a check on the lock is made, the HMS performs heartbeats on the lock record and the transaction record. It eventually ends up with the below errors if the number of jobs on the same table grew and commit at the same time. Ability to configure the delay between retries and slowing down retries further exponentially would help. Thanks.

```
MetaException(message:Unable to update transaction database org.postgresql.util.PSQLException: ERROR: could not serialize access due to read/write dependencies among transactions
Detail: Reason code: Canceled on identification as a pivot, during write.
Hint: The transaction might succeed if retried.
```

### CODE REVIEW

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 1]
```diff
@@ -70,21 +72,38 @@
 public class HiveTableOperations extends BaseMetastoreTableOperations {
   private static final Logger LOG = LoggerFactory.getLogger(HiveTableOperations.class);
 
-  private static final String HIVE_ACQUIRE_LOCK_STATE_TIMEOUT_MS = "iceberg.hive.lock-timeout-ms";
-  private static final long HIVE_ACQUIRE_LOCK_STATE_TIMEOUT_MS_DEFAULT = 3 * 60 * 1000; // 3 minutes
+  private static final String HIVE_ACQUIRE_LOCK_TIMEOUT_MS = "iceberg.hive.lock-timeout-ms";
+  private static final String HIVE_LOCK_CHECK_MIN_WAIT_MS = "iceberg.hive.lock-check-min-wait-ms";
+  private static final String HIVE_LOCK_CHECK_MAX_WAIT_MS = "iceberg.hive.lock-check-max-wait-ms";
+  private static final String HIVE_LOCK_CHECK_BACKOFF_SCALE_FACTOR = "iceberg.hive.lock-check-backoff-scale-factor";
+  private static final long HIVE_ACQUIRE_LOCK_TIMEOUT_MS_DEFAULT = 3 * 60 * 1000; // 3 minutes
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:79]: Might worth to mention in the documentation, or somewhere that this should be smaller than `hive.txn.timeout` or in newer versions `metastore.txn.timeout` otherwise the locks might be timed out without because of the lack of heartbeat.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 2]
```diff
@@ -70,21 +72,38 @@
...
+  private static final long HIVE_LOCK_CHECK_MIN_WAIT_MS_DEFAULT = 50; // 50 milliseconds
+  private static final long HIVE_LOCK_CHECK_MAX_WAIT_MS_DEFAULT = 5 * 1000; // 5 seconds
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: Might worth to mention in the documentation, or somewhere that this should be smaller than `hive.txn.timeout` or in newer versions `metastore.txn.timeout` otherwise the locks might be timed out without because of the lack of heartbeat.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java [hunk 3]
```diff
@@ -84,7 +84,7 @@
   private ExecutorService executorService;
   private TServer server;
   private HiveMetaStore.HMSHandler baseHandler;
-  private HiveClientPool clientPool;
+  protected HiveClientPool clientPool; // Exposed for testing.
```
Comment (pvary) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java:87]: nit: Could we use VisibleForTesting annotation here?

**Comment by pvary** [comment]
Minor comments, looks good to me (non-binding)

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (aokolnychyi) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: We should also add these configs to `configuration.md` to the rest of Hadoop conf, @raptond.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 4]
```diff
@@ -70,21 +72,38 @@
...
           String.class, String.class, Table.class, EnvironmentContext.class)
       .build();
 
+  protected static class WaitingForLockException extends RuntimeException {
+    public WaitingForLockException(String message) {
+      super(message);
+    }
+  }
+
+  private static final WaitingForLockException WAITING_FOR_LOCK_EXCEPTION =
```
Comment (aokolnychyi) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:96]: nit: it seems we use it only in one place. Would it make sense to just call `throw new` there?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 5]
```diff
@@ -70,21 +72,38 @@
...
+  private static final long HIVE_LOCK_CHECK_MAX_WAIT_MS_DEFAULT = 5 * 1000; // 5 seconds
+  private static final double HIVE_LOCK_CHECK_BACKOFF_SCALE_FACTOR_DEFAULT = 1.5;
   private static final DynMethods.UnboundMethod ALTER_TABLE = DynMethods.builder("alter_table")
       .impl(HiveMetaStoreClient.class, "alter_table_with_environmentContext",
           String.class, String.class, Table.class, EnvironmentContext.class)
       .impl(HiveMetaStoreClient.class, "alter_table",
           String.class, String.class, Table.class, EnvironmentContext.class)
       .build();
 
+  protected static class WaitingForLockException extends RuntimeException {
```
Comment (aokolnychyi) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:90]: Why protected?

Comment (aokolnychyi) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:90]: Why protected?

**Comment by aokolnychyi** [comment]
This change looks great to me, just minor comments in addition to what @pvary mentioned.
Thanks for working on this, @raptond!

We could add fewer configs but I'd be in favor of what this PR does. It was really painful when we hit this problem and couldn't do anything without changing the code so having props to configure every aspect sounds good to me. It is rather a sensitive area and more control here would be good.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 6]
```diff
@@ -304,32 +330,57 @@ private long acquireLock() throws UnknownHostException, TException, InterruptedE
...
-      lockResponse = metaClients.run(client -> client.checkLock(lockId));
-      state = lockResponse.getState();
 
-      // check timeout
-      duration = System.currentTimeMillis() - start;
-      if (duration > lockAcquireTimeout) {
+    if (state.get().equals(LockState.WAITING)) {
+      try {
+        Tasks.foreach(lockId)
+            .retry(Integer.MAX_VALUE - 100) // Endless retries bound by timeouts. Tasks.retry adds 1 for "first try".
```
Comment (RussellSpitzer) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:343]: Why - 100?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    doReturn(acquiredLockResponse).when(spyClient).lock(any());
+    doNothing().when(spyOps).doUnlock(eq(dummyLockId));
+
+    spyOps.doCommit(metadataV2, metadataV1);
+
+    Assert.assertEquals(1, spyOps.current().schema().columns().size()); // should be 1 again
+  }
+
+  @Test
+  public void testLockAcquisitionAfterRetries() throws TException, InterruptedException {
```
Comment (aokolnychyi) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java:120]: Is `InterruptedException` needed?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 8]
```diff
@@ -96,7 +115,14 @@ protected HiveTableOperations(Configuration conf, HiveClientPool metaClients, Fi
...
     this.lockAcquireTimeout =
-        conf.getLong(HIVE_ACQUIRE_LOCK_STATE_TIMEOUT_MS, HIVE_ACQUIRE_LOCK_STATE_TIMEOUT_MS_DEFAULT);
+        conf.getLong(HIVE_ACQUIRE_LOCK_TIMEOUT_MS, HIVE_ACQUIRE_LOCK_TIMEOUT_MS_DEFAULT);
+    this.lockCheckMinWaitTime =
+        conf.getLong(HIVE_LOCK_CHECK_MIN_WAIT_MS, HIVE_LOCK_CHECK_MIN_WAIT_MS_DEFAULT);
+    this.lockCheckMaxWaitTime =
+        conf.getLong(HIVE_LOCK_CHECK_MAX_WAIT_MS, HIVE_LOCK_CHECK_MAX_WAIT_MS_DEFAULT);
+    this.lockCheckBackoffScaleFactor =
+        conf.getDouble(HIVE_LOCK_CHECK_BACKOFF_SCALE_FACTOR, HIVE_LOCK_CHECK_BACKOFF_SCALE_FACTOR_DEFAULT);
+
```
Comment (aokolnychyi) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:125]: nit: extra empty line

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: The Glue catalog is introducing support for a lock using DynamoDB. It would be nice to standardize these options across catalogs so that we only need to document them once and they work the same way. FYI @jackye1995.

I think it would also make sense for these to be catalog options, rather than pulled from the Hive configuration. We used HiveConf originally because we didn't have catalog-specific configuration, but now I think it would make sense to move these into catalog properties. We don't want to increase the cases where we use a Hadoop Configuration.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 4 above)
Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:96]: I agree. What is the purpose of having a single instance of an exception? Wouldn't that lead to deceptive stack traces?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 9]
```diff
@@ -96,7 +115,14 @@ protected HiveTableOperations(Configuration conf, HiveClientPool metaClients, Fi
...
     this.tableName = table;
     this.lockAcquireTimeout =
-        conf.getLong(HIVE_ACQUIRE_LOCK_STATE_TIMEOUT_MS, HIVE_ACQUIRE_LOCK_STATE_TIMEOUT_MS_DEFAULT);
+        conf.getLong(HIVE_ACQUIRE_LOCK_TIMEOUT_MS, HIVE_ACQUIRE_LOCK_TIMEOUT_MS_DEFAULT);
+    this.lockCheckMinWaitTime =
+        conf.getLong(HIVE_LOCK_CHECK_MIN_WAIT_MS, HIVE_LOCK_CHECK_MIN_WAIT_MS_DEFAULT);
+    this.lockCheckMaxWaitTime =
+        conf.getLong(HIVE_LOCK_CHECK_MAX_WAIT_MS, HIVE_LOCK_CHECK_MAX_WAIT_MS_DEFAULT);
+    this.lockCheckBackoffScaleFactor =
+        conf.getDouble(HIVE_LOCK_CHECK_BACKOFF_SCALE_FACTOR, HIVE_LOCK_CHECK_BACKOFF_SCALE_FACTOR_DEFAULT);
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:124]: We don't expose the scale factor in other places. I think it makes sense to keep it simple and not have one here either.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 10]
```diff
@@ -304,32 +330,57 @@ private long acquireLock() throws UnknownHostException, TException, InterruptedE
...
+            .onlyRetryOn(WaitingForLockException.class)
+            .run(id -> {
+              try {
+                LockResponse response = metaClients.run(client -> client.checkLock(id));
+                LockState newState = response.getState();
+                state.set(newState);
+                if (newState.equals(LockState.WAITING)) {
+                  throw WAITING_FOR_LOCK_EXCEPTION;
+                }
+              } catch (InterruptedException | TException e) {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:359]: For `InterruptedException`, why not throw `WaitingForLockException` and signal that the thread was interrupted? Then this could use the checked exception call, `run(id -> {...}, TException.class)` and would not need to wrap the exceptions.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 5 above)
Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:90]: I agree. If this is always caught, it should be private.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (jackye1995) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: Yes agree. But Hive is currently built around reading from Hadoop configs. If we want to change it to use catalog properties, we need to also change all the places that loads `HiveCatalog` using the constructor `public HiveCatalog(Configuration conf)`, such as https://github.com/apache/iceberg/blob/master/mr/src/main/java/org/apache/iceberg/mr/Catalogs.java#L215

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 4 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:96]: It's not needed to be a single instance and holding a different stack trace than where it originated. I'll fix this.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 5 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:90]: It's a miss from using in a testing. Fixing it.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 9 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:124]: I'll remove the scale factor.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (aokolnychyi) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: This API is also used from Spark 2 where we don't have a way to specify catalog options. Plus, we already have a Hadoop conf for the lock timeout. How do we approach this?

Comment (aokolnychyi) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: That being said, I am +1 for adding catalog options.

Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:81]: In that case, we should default from Configuration, but prefer options passed to the initialize method.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 10 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: The code looks better after this comment. However throwing `WaitingForLockException` ends up losing the source of the original `InterruptedException` because it would get handled here: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/util/Tasks.java#L452

So, I chose to throw `RuntimeException` which will stop the retry and preserve the source stack trace.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 6 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:326]: I only wanted to keep a big number for retry.  Eg `Integer.MAX_VALUE`. But, the [setter](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/util/Tasks.java#L163) adds 1 overflowing to `MIN_VALUE`. 

`Integer.MAX_VALUE - 1` would simply suffice, but I chose conservatively to set `Integer.MAX_VALUE - 100`.

Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:326]: I think it would be worth noting the rational for a choice like this in a comment.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 11]
```diff
@@ -66,33 +68,50 @@
...
   private final String tableName;
   private final Configuration conf;
   private final long lockAcquireTimeout;
+  private final long lockCheckMinWaitTime;
+  private final long lockCheckMaxWaitTime;
   private final FileIO fileIO;
 
   protected HiveTableOperations(Configuration conf, HiveClientPool metaClients, FileIO fileIO,
-                                String catalogName, String database, String table) {
+      String catalogName, String database, String table) {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:101]: Nit: unnecessary whitespace change.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 12]
```diff
@@ -294,32 +313,50 @@ private long acquireLock() throws UnknownHostException, TException, InterruptedE
...
+              try {
+                LockResponse response = metaClients.run(client -> client.checkLock(id));
+                LockState newState = response.getState();
+                state.set(newState);
+                if (newState.equals(LockState.WAITING)) {
+                  throw new WaitingForLockException("Waiting for lock.");
+                }
+              } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException("Interrupted while checking lock status.", e);
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:344]: I don't think it is necessary to throw `RuntimeException` here. If this doesn't throw `WaitingForLockException` then it will exit and move on. Since timeout is not set, it would hit the check for whether the lock was acquired and fail, resulting in the `CommitFailedException`.

I think that's a fairly reasonable way to handle an interrupt without wrapping it in a `RuntimeException`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 10 above)
Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: I'd probably opt to suppress the interrupt and let the code carry on after setting that the thread was interrupted. That results in a `CommitFailedException`. I don't think that preserving the stack of the `InterruptedException` is really needed, but I'm fin with it this way if you prefer it.

**Review by rdblue — APPROVED** [review]
Overall looks good to me. @aokolnychyi can you take another look?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java (hunk 7 above)
Comment (raptond) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java:121]: yes, here the [HiveTableOperations.doUnlock](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java#L340) method throws `InterruptedException`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 6 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: 👍 +1 I have added the rationale in the comments.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 8 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: taken care.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 12 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: You are correct. I was fixated on failing the execution. This suggestion works nicely and I have a test case for this. Thank you.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 10 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: Taken care as per [the other comment](https://github.com/apache/iceberg/pull/1873#discussion_r539077360).

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 3 above)
Comment (raptond) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Since this was a test class, I didn't add this annotation. I finally ended up not using it, so I reverted back the change. Thanks for the review.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:76]: @aokolnychyi @pvary - 
image

**Comment by raptond** [comment]
All comments taken care.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 11 above)
Comment (raptond) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: done.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 2 above)
Comment (aokolnychyi) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:76]: To make sure I got @rdblue and @jackye1995. You are talking about generalizing catalog options, right?

**Comment by aokolnychyi** [comment]
I'll do a pass in 15 mins

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks everyone! The change looks solid so I merged it!

**Comment by aokolnychyi** [comment]
@raptond, do you want to work on catalog options for this next?

**Comment by raptond** [comment]
@aokolnychyi - Sure, I will work on to submit a new one with catalog options (properties via initialize method) overriding the Configuration.
